### ISSUE TITLE
Include non indexed repos in StatusMessages GraphQL query

### ISSUE THREAD

**Issue by ryanslade** [sourcegraph/sourcegraph#18543]
This is needed to support the new notifications system, see [Figma](https://www.figma.com/file/0567AnXEpQTzQXkgs7AzEh/%2316510-%E2%80%93-WIP-%E2%80%93-Status-notification-system?node-id=466%3A279)

This issue focuses only on supporting users, not orgs.

### Current status 

We have the `StatusMessages` query, defined as:

```graphql
"""
FOR INTERNAL USE ONLY: Lists all status messages
"""
statusMessages: [StatusMessage!]!

"""
FOR INTERNAL USE ONLY: A status message
"""
union StatusMessage = CloningProgress | ExternalServiceSyncError | SyncError

"""
FOR INTERNAL USE ONLY: A status message produced when repositories are being
cloned
"""
type CloningProgress {
    """
    The message of this status message
    """
    message: String!
}

"""
FOR INTERNAL USE ONLY: A status message produced when repositories could not
be synced from an external service
"""
type ExternalServiceSyncError {
    """
    The message of this status message
    """
    message: String!
    """
    The external service that failed to sync
    """
    externalService: ExternalService!
}

"""
FOR INTERNAL USE ONLY: A status message produced when repositories could not
be synced
"""
type SyncError {
    """
    The message of this status message
    """
    message: String!
}
``` 

The actual implementation does the following:

* CloningProgress simply returns the number of uncloned repos
* ExternalServiceSyncError returns errors associated with each external service
* SyncError is never returned

The results are scoped to the user making the request: 

* For an ordinary user it only shows their own external services and uncloned repos they've added
* For a site admin, it shows their own external services and any other site level external services. It shows ALL uncloned repos for the entire site. On Cloud we do not list this as repos can become uncloned due to cleanup processes. 

### What needs to be added

* The number of repos that have been cloned, but have not been indexed at all. This should be scoped in the same way as uncloned repos above.
* The number of repos that encountered an error during their last sync. This is different to a metadata sync to the code host which is what `ExternalServiceSyncError` indicates. 

@keegancsmith @tsenart I'm not sure what the best approach is to add these fields, can we sync on this? 

@quinnkeast Am I correct in assuming that we only care about first clone and first index?

**Comment by keegancsmith** [comment]
> The number of repos that have been cloned, but have not been indexed at all. This should be scoped in the same way as uncloned repos above.

You have fast access to `map[string]*zoekt.Repository` of indexed repositories (grep for ListAll). So to compute you would have to diff in memory. Tomas has mentioned this a few times, but we might want to sync indexed state to the DB to make this fast since it is just a count.

> The number of repos that encountered an error during their last sync. This is different to a metadata sync to the code host which is what ExternalServiceSyncError indicates.

Do we store this anywhere? I assume we would need a new table in the DB for this sort of info?

**Comment by quinnkeast** [comment]
>@quinnkeast Am I correct in assuming that we only care about first clone and first index?

I'm going back and forth on this. The first clone and index is the most important, but this is also meant to create confidence Sourcegraph accurately reflects the latest code. Is there a scenario where it might take a long time to bring an existing repository up to date?

**Comment by ryanslade** [comment]
> Do we store this anywhere? I assume we would need a new table in the DB for this sort of info?

@keegancsmith No, I was thinking a new column in the repo table? `sync_error` which would store the error of the most recent sync or set it to `null` if there was no error. But, would we not need to make the update from `gitserver`, how do you feel about that?

> Is there a scenario where it might take a long time to bring an existing repository up to date?

@quinnkeast If a repo hasn't been modified in a while and a change is pushed to it, it could take while to reflect in Sourcegraph. We back off for up to 8 hours depending on how frequently a repos changes.

https://github.com/sourcegraph/sourcegraph/blob/c59ea3c27d6c3130fdeee3e64e76d0ed184a2b4c/internal/repos/scheduler.go#L89-L93

**Comment by keegancsmith** [comment]
> > Do we store this anywhere? I assume we would need a new table in the DB for this sort of info?
> 
> @keegancsmith No, I was thinking a new column in the repo table? `sync_error` which would store the error of the most recent sync or set it to `null` if there was no error. But, would we not need to make the update from `gitserver`, how do you feel about that?

Seems fine to write from the gitserver process. Why not a new table? The repo table is going to grow into a monster if we just keep tacking on new columns.

**Comment by ryanslade** [comment]
@keegancsmith I proposed a new table [here](https://github.com/sourcegraph/sourcegraph/pull/18692), which is pretty similar to what you and @tsenart prototyped a while back.
